### ISSUE TITLE
[mingw-w64-headers] why use ${MINGW_PREFIX}/${MINGW_CHOST} prefix ?

### ISSUE THREAD

**Issue by elmarco** [msys2/MINGW-packages#10630]
Hi,

On Fedora, mingw-headers is installed under the regular target sysroot prefix (for ex `/usr/x86_64-w64-mingw32/sys-root/mingw/`) with the rest of the packages and libraries.

Why is it using a special prefix in msys2?

(fwiw, I am doing an experiment trying to replace the mingw-* packages with a msys2 installation on Linux and bwrap environment, it seems to be doable. has anyone tried that before? that would avoid the proliferation of mingw* packaging and duplication of effort in various distributions, and we could all use msys2 eventually :)

**Comment by elmarco** [comment]
Same question for mingw-w64-crt.

**Comment by Biswa96** [comment]
mingw packages use different prefix than msys2 packages. For mingw packages, it is $MINGW_PREFIX (e.g. mingw64). But /usr is for msys2 packages.

**Comment by elmarco** [comment]
> mingw packages use different prefix than msys2 packages. For mingw packages, it is $MINGW_PREFIX (e.g. mingw64). But /usr is for msys2 packages.

yes, that's not my point here.

Under msys2, for mingw64 target, mingw-headers & mingw-crt are installed in /mingw64/x86_64-w64-mingw32, while the rest of the packages are under /mingw64.

(similarly on Fedora all packages, including crt & headers, are installed in `/usr/x86_64-w64-mingw32/sys-root/mingw`)

**Comment by 1480c1** [comment]
Looking at the history of the PKGBUILD for gcc, it seems the first spotting of `${MINGW_PREFIX}/${MINGW_CHOST}` is https://github.com/msys2/MINGW-packages/commit/26344720d82a575f7c59fa0ac65ae4ff3e3fc219

and https://github.com/msys2/MINGW-packages/commit/fd87286156cc41c693c0d44f6a1a7722f051b1d7 seems to be related as well

It seems to be a historical thing(?)

**Comment by mati865** [comment]
I can only guess but I think it was done to separate different subsystems from each other.
Right now `/usr` is stuff based on Cygwin (uses POSIX paths and is slow because of emulation), the other subsystems are native Windows applications so their binaries are strongly preferred. To keep that separation one would need to create non-standard `/usr/x86_64-w64-mingw32/bin` anyway and put native executables there.

**Comment by HolyBlackCat** [comment]
> msys2 installation on Linux ... has anyone tried that before?

Yes, but without `bwrap` (yet): https://github.com/HolyBlackCat/quasi-msys2

**Comment by elmarco** [comment]
> > msys2 installation on Linux ... has anyone tried that before?
> 
> Yes, but without `bwrap` (yet): https://github.com/HolyBlackCat/quasi-msys2

That's way more complicated than what I am looking for. On fedora, we have cross-compilers and various dev tools or macros to cross compile. pacman can be installed too. So all that would be needed pretty much is to replace the mingw/sys-root with the one from msys2 (assuming the compilers are compatible). I have no need for wine, binfmt etc. I am playing with a script like that: https://gist.github.com/elmarco/0178036a288d0859567a5a4f88409237. I am waiting for this bug to be fixed to actually run more checks. Your experience could be really valuable, but I why not simply use a container/namespace with pacman?

**Comment by HolyBlackCat** [comment]
@elmarco
> why not simply use a container/namespace with pacman?

I wasn't sure how to point pacman to MSYS2 repos, or if it would work at all (since MSYS uses a patched pacman). Writing a small custom package manager seemed more straightforward.

Neither Wine nor binfmt_misc are strictly necessary for it to function. They're used to help convince CMake & Autotools that we're not cross-compiling, but are otherwise unnecessary.

Clang was able to cross-compile from MSYS2 MINGW64 sysroot with minimal configuration. Just `--target` and `--sysroot`, I believe.
